File: src/main/java/com/pedigree/model/GedcomAttribute.java
package com.pedigree.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.UUID;

/**
 * Represents a generic GEDCOM attribute (tag-value pair with optional sub-attributes).
 * Used for custom tags or attributes like PEDI, AGE, CAUS, etc.
 */
public class GedcomAttribute {
    private final String id;
    private String tag;
    private String value;

    public GedcomAttribute() {
        this(UUID.randomUUID().toString());
    }

    public GedcomAttribute(String id) {
        this.id = id;
    }

    @JsonCreator
    public GedcomAttribute(
            @JsonProperty("id") String id,
            @JsonProperty("tag") String tag,
            @JsonProperty("value") String value
    ) {
        this.id = id != null ? id : UUID.randomUUID().toString();
        this.tag = tag;
        this.value = value;
    }

    public String getId() { return id; }
    public String getTag() { return tag; }
    public void setTag(String tag) { this.tag = tag; com.pedigree.util.DirtyFlag.setModified(); }
    public String getValue() { return value; }
    public void setValue(String value) { this.value = value; com.pedigree.util.DirtyFlag.setModified(); }
}

File: src/main/kotlin/com/family/tree/model/GedcomAttribute.kt
package com.family.tree.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.family.tree.util.DirtyFlag
import java.util.UUID

/**
 * Represents a generic GEDCOM attribute (tag-value pair with optional sub-attributes).
 * Used for custom tags or attributes like PEDI, AGE, CAUS, etc.
 */
class GedcomAttribute {
    val id: String
    var tag: String? = null
        set(value) { field = value; DirtyFlag.setModified() }
    var value: String? = null
        set(value) { field = value; DirtyFlag.setModified() }

    constructor() : this(UUID.randomUUID().toString())

    constructor(id: String) {
        this.id = id
    }

    @JsonCreator
    constructor(
        @JsonProperty("id") id: String?,
        @JsonProperty("tag") tag: String?,
        @JsonProperty("value") value: String?
    ) {
        this.id = id ?: UUID.randomUUID().toString()
        this.tag = tag
        this.value = value
    }
}
